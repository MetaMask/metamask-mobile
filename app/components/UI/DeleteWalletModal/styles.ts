/* eslint-disable import/prefer-default-export */
import { fontStyles } from '../../../styles/common';
import { StyleSheet } from 'react-native';
import Device from '../../../util/device';

const breakPoint = Device.getDeviceHeight() < 700;

export const createStyles = (colors: any) =>
  StyleSheet.create({
    bottomModal: {
      justifyContent: 'flex-end',
      margin: 0,
    },
    areYouSure: {
      width: '100%',
      padding: breakPoint ? 16 : 24,
      justifyContent: 'center',
      alignSelf: 'center',
    },
    warningIcon: {
      alignSelf: 'center',
      color: colors.error.default,
      marginVertical: 10,
    },
    heading: {
      marginHorizontal: 6,
      color: colors.text.default,
      ...fontStyles.bold,
      fontSize: 20,
      textAlign: 'center',
      lineHeight: breakPoint ? 24 : 26,
    },
    red: {
      marginHorizontal: 24,
      color: colors.error.default,
    },
    warningText: {
      ...fontStyles.normal,
      textAlign: 'center',
      fontSize: 14,
      lineHeight: breakPoint ? 18 : 22,
      color: colors.text.default,
      marginTop: 20,
    },
    bold: {
      ...fontStyles.bold,
    },
    delete: {
      marginBottom: 20,
    },
    input: {
      ...fontStyles.normal,
      fontSize: 16,
      paddingTop: 2,
      color: colors.text.default,
    },
    deleteWarningMsg: {
      ...fontStyles.normal,
      fontSize: 16,
      lineHeight: 20,
      marginTop: 10,
      color: colors.error.default,
    },
  });
