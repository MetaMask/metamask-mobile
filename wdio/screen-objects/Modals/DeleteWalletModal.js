import Gestures from '../../helpers/Gestures';
import Selectors from '../../helpers/Selectors';
import {
  DELETE_MODAL_UNDERSTAND_CONTINUE_ID,
  DELETE_MODAL_DELETE_INPUT_ID,
  DELETE_MODAL_DELETE_MY_WALLET_PERMANENTLY,
} from '../testIDs/Components/DeleteWalletModal.testIds';

class DeleteWalletModal {
  get iUnderstandContinueButton() {
    return Selectors.getElementByPlatform(DELETE_MODAL_UNDERSTAND_CONTINUE_ID);
  }

  get deleteInput() {
    return Selectors.getElementByPlatform(DELETE_MODAL_DELETE_INPUT_ID);
  }

  get deleteMyWalletButton() {
    return Selectors.getElementByPlatform(
      DELETE_MODAL_DELETE_MY_WALLET_PERMANENTLY,
    );
  }

  async tapIUnderstandContinue(text) {
    await Gestures.waitAndTap(this.iUnderstandContinueButton);
  }

  async typeTextDelete(deleteText) {
    await Gestures.typeText(this.deleteInput, deleteText, false);
  }

  async tapDeleteMyWallet() {
    await Gestures.waitAndTap(this.deleteMyWalletButton);
  }
}

export default new DeleteWalletModal();
