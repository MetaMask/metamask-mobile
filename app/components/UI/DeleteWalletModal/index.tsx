import React, { useState, useRef } from 'react';
import {
  View,
  Text,
  TouchableWithoutFeedback,
  Keyboard,
  InteractionManager,
  UIManager,
  LayoutAnimation,
  Platform,
} from 'react-native';
import { useNavigation } from '@react-navigation/native';
import Icon from 'react-native-vector-icons/FontAwesome';
import { OutlinedTextField } from 'react-native-material-textfield';
import { createStyles } from './styles';
import ReusableModal, { ReusableModalRef } from '../ReusableModal';
import WarningExistingUserModal from '../WarningExistingUserModal';
import { useDeleteWallet } from '../../hooks/DeleteWallet';
import { strings } from '../../../../locales/i18n';
import {
  DELETE_WALLET_CONTAINER_ID,
  DELETE_WALLET_INPUT_BOX_ID,
} from '../../../constants/test-ids';
import { tlc } from '../../../util/general';
import { useTheme } from '../../../util/theme';
import Device from '../../../util/device';
import Routes from '../../../constants/navigation/Routes';
import {
  DELETE_MODAL_UNDERSTAND_CONTINUE_ID,
  DELETE_MODAL_CANCEL_BUTTON,
  DELETE_MODAL_DELETE_MY_WALLET_PERMANENTLY,
  DELETE_MODEL_DELETE_MY_WALLET_CANCEL,
} from '../../../../wdio/screen-objects/testIDs/Components/DeleteWalletModal.testIds';
import generateTestId from '../../../../wdio/utils/generateTestId';

const DELETE_KEYWORD = 'delete';

if (Device.isAndroid() && UIManager.setLayoutAnimationEnabledExperimental) {
  UIManager.setLayoutAnimationEnabledExperimental(true);
}

const DeleteWalletModal = () => {
  const navigation = useNavigation();
  const { colors, themeAppearance } = useTheme();
  const styles = createStyles(colors);

  const modalRef = useRef<ReusableModalRef>(null);

  const [showConfirm, setShowConfirm] = useState<boolean>(false);
  const [deleteText, setDeleteText] = useState<string>('');
  const [disableButton, setDisableButton] = useState<boolean>(true);

  const [resetWalletState, deleteUser] = useDeleteWallet();

  const showConfirmModal = () => {
    setShowConfirm(true);
    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
  };

  const isTextDelete = (text: string) => tlc(text) === DELETE_KEYWORD;

  const checkDelete = (text: string) => {
    setDeleteText(text);
    setDisableButton(!isTextDelete(text));
  };

  const dismissModal = (cb?: () => void): void =>
    modalRef?.current?.dismissModal(cb);

  const triggerClose = (): void => dismissModal();

  const navigateOnboardingRoot = (): void => {
    navigation.reset({
      routes: [
        {
          name: Routes.ONBOARDING.ROOT_NAV,
          state: {
            routes: [
              {
                name: Routes.ONBOARDING.NAV,
                params: {
                  screen: Routes.ONBOARDING.ONBOARDING,
                  params: { delete: true },
                },
              },
            ],
          },
        },
      ],
    });
  };

  const deleteWallet = async () => {
    triggerClose();
    await resetWalletState();
    await deleteUser();
    InteractionManager.runAfterInteractions(() => {
      navigateOnboardingRoot();
    });
  };

  return (
    <ReusableModal ref={modalRef}>
      {showConfirm ? (
        <WarningExistingUserModal
          warningModalVisible
          cancelText={strings('login.delete_my')}
          cancelTestID={DELETE_MODAL_DELETE_MY_WALLET_PERMANENTLY}
          confirmTestID={DELETE_MODEL_DELETE_MY_WALLET_CANCEL}
          cancelButtonDisabled={disableButton}
          onCancelPress={deleteWallet}
          onRequestClose={triggerClose}
          onConfirmPress={triggerClose}
        >
          <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
            <View style={styles.areYouSure}>
              <Text style={[styles.heading, styles.delete]}>
                {strings('login.type_delete', {
                  [DELETE_KEYWORD]: DELETE_KEYWORD,
                })}
              </Text>
              <OutlinedTextField
                style={styles.input}
                testID={DELETE_WALLET_INPUT_BOX_ID}
                {...generateTestId(Platform, DELETE_WALLET_INPUT_BOX_ID)}
                autoFocus
                returnKeyType={'done'}
                onChangeText={checkDelete}
                autoCapitalize="none"
                value={deleteText}
                baseColor={colors.border.default}
                tintColor={colors.primary.default}
                placeholderTextColor={colors.text.muted}
                keyboardAppearance={themeAppearance}
              />
            </View>
          </TouchableWithoutFeedback>
        </WarningExistingUserModal>
      ) : (
        <WarningExistingUserModal
          warningModalVisible
          cancelText={strings('login.i_understand')}
          onCancelPress={showConfirmModal}
          onRequestClose={triggerClose}
          onConfirmPress={triggerClose}
          cancelTestID={DELETE_MODAL_UNDERSTAND_CONTINUE_ID}
          confirmTestID={DELETE_MODAL_CANCEL_BUTTON}
        >
          <View style={styles.areYouSure} testID={DELETE_WALLET_CONTAINER_ID}>
            <Icon
              style={styles.warningIcon}
              size={46}
              color={colors.error.default}
              name="exclamation-triangle"
            />
            <Text style={[styles.heading, styles.red]}>
              {strings('login.are_you_sure')}
            </Text>
            <Text style={styles.warningText}>
              <Text>{strings('login.your_current_wallet')}</Text>
              <Text style={styles.bold}>{strings('login.removed_from')}</Text>
              <Text>{strings('login.this_action')}</Text>
            </Text>
            <Text style={[styles.warningText]}>
              <Text>{strings('login.you_can_only')}</Text>
              <Text style={styles.bold}>
                {strings('login.recovery_phrase')}
              </Text>
              <Text>{strings('login.metamask_does_not')}</Text>
            </Text>
          </View>
        </WarningExistingUserModal>
      )}
    </ReusableModal>
  );
};

export default React.memo(DeleteWalletModal);
