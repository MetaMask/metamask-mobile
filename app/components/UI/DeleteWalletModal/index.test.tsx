import React from 'react';
import DeleteWalletModal from './';
import renderWithProvider from '../../../util/test/renderWithProvider';
import { backgroundState } from '../../../util/test/initial-root-state';

const mockInitialState = {
  engine: { backgroundState },
  security: {
    dataCollectionForMarketing: false,
  },
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => jest.fn(),
  useSelector: jest.fn(),
}));
const mockNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
    }),
  };
});

describe('DeleteWalletModal', () => {
  it('should render correctly', () => {
    const wrapper = renderWithProvider(<DeleteWalletModal />, {
      state: mockInitialState,
    });

    expect(wrapper).toMatchSnapshot();
  });
});
