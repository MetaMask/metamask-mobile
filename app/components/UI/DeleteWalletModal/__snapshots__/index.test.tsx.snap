// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteWalletModal should render correctly 1`] = `
<View
  style={
    {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#00000099",
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "opacity": undefined,
        },
      ]
    }
  />
  <View
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      [
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          {
            "transform": [
              {
                "translateY": 1334,
              },
            ],
          },
        ],
        undefined,
      ]
    }
  >
    <TouchableOpacity
      disabled={false}
      onPress={[Function]}
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <Modal
      animationType="none"
      deviceHeight={null}
      deviceWidth={null}
      hardwareAccelerated={false}
      hideModalContentWhileAnimating={false}
      onBackdropPress={[Function]}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onRequestClose={[Function]}
      onSwipeComplete={[Function]}
      panResponderThreshold={4}
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      statusBarTranslucent={false}
      supportedOrientations={
        [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection="down"
      swipeThreshold={100}
      transparent={true}
      visible={true}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#00000099",
            "bottom": 0,
            "height": 1334,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "width": 750,
          }
        }
      />
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          [
            [
              {
                "margin": 37.5,
                "transform": [
                  {
                    "translateY": 0,
                  },
                ],
              },
              {
                "flex": 1,
                "justifyContent": "center",
              },
              [
                {
                  "margin": 0,
                  "width": "100%",
                },
                undefined,
              ],
              {
                "margin": 0,
              },
            ],
            {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <View
          collapsable={false}
          deviceHeight={null}
          deviceWidth={null}
          hideModalContentWhileAnimating={false}
          onBackdropPress={[Function]}
          onModalHide={[Function]}
          onModalWillHide={[Function]}
          onModalWillShow={[Function]}
          onMoveShouldSetResponder={[Function]}
          onMoveShouldSetResponderCapture={[Function]}
          onResponderEnd={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderReject={[Function]}
          onResponderRelease={[Function]}
          onResponderStart={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onStartShouldSetResponderCapture={[Function]}
          onSwipeComplete={[Function]}
          panResponderThreshold={4}
          pointerEvents="box-none"
          scrollHorizontal={false}
          scrollOffset={0}
          scrollOffsetMax={0}
          scrollTo={null}
          statusBarTranslucent={false}
          style={
            {
              "flex": 1,
              "justifyContent": "center",
              "left": 0,
              "margin": 0,
              "top": 0,
              "transform": [
                {
                  "translateY": 1334,
                },
              ],
              "width": "100%",
            }
          }
          supportedOrientations={
            [
              "portrait",
              "landscape",
            ]
          }
          swipeDirection="down"
          swipeThreshold={100}
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "column",
                  "justifyContent": "center",
                  "marginHorizontal": 24,
                },
                null,
              ]
            }
          >
            <View
              style={
                [
                  {
                    "backgroundColor": "#ffffff",
                    "borderRadius": 10,
                    "width": "100%",
                  },
                  null,
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                    },
                    null,
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignSelf": "center",
                      "justifyContent": "center",
                      "padding": 24,
                      "width": "100%",
                    }
                  }
                  testID="delete-wallet-modal-container"
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      [
                        {
                          "color": "#d73847",
                          "fontSize": 46,
                        },
                        {
                          "alignSelf": "center",
                          "color": "#d73847",
                          "marginVertical": 10,
                        },
                        {
                          "fontFamily": "FontAwesome",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    ÔÅ±
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#141618",
                          "fontFamily": "EuclidCircularB-Bold",
                          "fontSize": 20,
                          "fontWeight": "600",
                          "lineHeight": 26,
                          "marginHorizontal": 6,
                          "textAlign": "center",
                        },
                        {
                          "color": "#d73847",
                          "marginHorizontal": 24,
                        },
                      ]
                    }
                  >
                    Are you sure you want to erase your wallet?
                  </Text>
                  <Text
                    style={
                      {
                        "color": "#141618",
                        "fontFamily": "EuclidCircularB-Regular",
                        "fontSize": 14,
                        "fontWeight": "400",
                        "lineHeight": 22,
                        "marginTop": 20,
                        "textAlign": "center",
                      }
                    }
                  >
                    <Text>
                      Your current wallet, accounts and assets will be 
                    </Text>
                    <Text
                      style={
                        {
                          "fontFamily": "EuclidCircularB-Bold",
                          "fontWeight": "600",
                        }
                      }
                    >
                      removed from this app permanently. 
                    </Text>
                    <Text>
                      This action cannot be undone.
                    </Text>
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#141618",
                          "fontFamily": "EuclidCircularB-Regular",
                          "fontSize": 14,
                          "fontWeight": "400",
                          "lineHeight": 22,
                          "marginTop": 20,
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    <Text>
                      You can ONLY recover this wallet with your 
                    </Text>
                    <Text
                      style={
                        {
                          "fontFamily": "EuclidCircularB-Bold",
                          "fontWeight": "600",
                        }
                      }
                    >
                      Secret Recovery Phrase 
                    </Text>
                    <Text>
                      MetaMask does not have your Secret Recovery Phrase.
                    </Text>
                  </Text>
                </View>
              </View>
              <View
                style={
                  [
                    {
                      "flexDirection": "column",
                      "paddingHorizontal": 16,
                      "paddingVertical": 8,
                    },
                    undefined,
                  ]
                }
              >
                <TouchableOpacity
                  accessibilityRole="button"
                  accessible={true}
                  activeOpacity={0.2}
                  disabled={false}
                  onPress={[Function]}
                  style={
                    [
                      [
                        {
                          "borderRadius": 100,
                          "justifyContent": "center",
                          "padding": 15,
                        },
                        {
                          "backgroundColor": "#d73847",
                        },
                        [
                          {
                            "margin": 8,
                          },
                          false,
                        ],
                      ],
                      null,
                    ]
                  }
                  testID="delete-modal-understand-continue"
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#007aff",
                          "fontSize": 17,
                          "fontWeight": "500",
                          "textAlign": "center",
                        },
                        null,
                        [
                          {
                            "fontFamily": "EuclidCircularB-Bold",
                            "fontSize": 14,
                            "fontWeight": "600",
                            "textAlign": "center",
                          },
                          {
                            "color": "#ffffff",
                          },
                          undefined,
                        ],
                        null,
                      ]
                    }
                  >
                    I understand, continue
                  </Text>
                </TouchableOpacity>
                <TouchableOpacity
                  accessibilityRole="button"
                  accessible={true}
                  activeOpacity={0.2}
                  disabled={false}
                  onPress={[Function]}
                  style={
                    [
                      [
                        {
                          "borderRadius": 100,
                          "justifyContent": "center",
                          "padding": 15,
                        },
                        {
                          "backgroundColor": "#ffffff",
                          "borderColor": "#6a737d",
                          "borderWidth": 1,
                        },
                        [
                          {
                            "margin": 8,
                          },
                          false,
                        ],
                      ],
                      null,
                    ]
                  }
                  testID="delete-modal-cancel-button"
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#007aff",
                          "fontSize": 17,
                          "fontWeight": "500",
                          "textAlign": "center",
                        },
                        null,
                        [
                          {
                            "fontFamily": "EuclidCircularB-Bold",
                            "fontSize": 14,
                            "fontWeight": "600",
                            "textAlign": "center",
                          },
                          {
                            "color": "#6a737d",
                          },
                          undefined,
                        ],
                        null,
                      ]
                    }
                  >
                    Cancel
                  </Text>
                </TouchableOpacity>
              </View>
            </View>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</View>
`;
